soccer/planning/planner/LineKickPlanner.cpp

```diff
@@ -0,0 +1,204 @@
+        }
+    }
+
+    // The ball is moving slowly so we can pretty much treat it as stationary
+    if (ball.vel.mag() < 0.2) {
+        return planForSlowMovingBall(std::move(request));
+    }
+
+    if (!prevTrajectory.empty() && _targetKickPos) {
+        RJ::Seconds timeInto = RJ::now() - prevTrajectory.begin_time();
+        RJ::Seconds timeLeft = prevTrajectory.duration() - timeInto;
+
+        Point targetPos;
+        RJ::Time time = ball.estimateTimeTo(*_targetKickPos, &targetPos);
+        RJ::Seconds timeToHit = time - curTime;
+        if (timeLeft < RJ::Seconds(1.0)) {
+            Point targetVel =
+                (aimTarget - targetPos).normalized(*_approachSpeed);
+            targetPos -= targetVel.normalized(Robot_Radius + Ball_Radius * 2);
+
+            _finalApproach = true;
+            RobotInstant goalInstant{Pose{targetPos, 0}, Twist{targetVel, 0},
+                                     RJ::Time{0s}};
+            Trajectory path = CreatePath::simple(startInstant, goalInstant,
+                                                 request.constraints.mot);
+            if (!path.empty()) {
+                path.setDebugText(
+                    "FinalPath" + QString::number(path.duration().count()) +
+                    " " + QString::number(timeToHit.count()) + " " +
+                    QString::number(time.time_since_epoch().count()));
+                PlanAngles(path, startInstant, AngleFns::facePoint(aimTarget),
+                           rotationConstraints);
+                return std::move(path);
+            }
+        }
+
+        if (!prevTrajectory.hit(request.static_obstacles, timeInto) &&
+            timeLeft - 1000ms < timeToHit && _reusePathCount < 10) {
+            _reusePathCount++;
+            Point nearPoint;
+            prevTrajectory.setDebugText("Reuse prevPath");
+            if (ball.estimateTimeTo(prevTrajectory.last().pose.position(),
+                                    &nearPoint) >=
+                RJ::now() + timeLeft - 1000ms) {
+                return reuse(std::move(request));
+            }
+        }
+    }
+
+    std::optional<Trajectory> brutePath = attemptBruteForce(request);
+    if (brutePath) return std::move(*brutePath);
+
+    Point targetVel = (aimTarget - ball.pos).normalized(*_approachSpeed);
+    Point targetPos = ball.pos - targetVel.normalized(Robot_Radius * 3);
+
+    request.dynamic_obstacles.push_back(ball.dynamicObs());
+    RobotInstant targetInstant{Pose{targetPos, 0}, Twist{targetVel, 0},
+                               RJ::Time{0s}};
+    request.motionCommand = PathTargetCommand{targetInstant};
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    path.setDebugText("Gives ups");
+    _targetKickPos = std::nullopt;
+    PlanAngles(path, startInstant, AngleFns::facePoint(aimTarget),
+               rotationConstraints);
+    return std::move(path);
+}
+
+Trajectory LineKickPlanner::planForSlowMovingBall(PlanRequest&& request) {
```

Comment:
Does line kick even work on a fast moving ball?

Reply:
lol, it barely even used to work on slow moving balls